compose/utils.py

```diff
@@ -80,6 +72,36 @@ def inner_execute_function(an_callable, parameter, msg_index):
             raise error
 
 
+class ParallelStreamWriter(object):
+    """Write out messages for operations happening in parallel.
+
+    Each operation has it's own line, and ANSI code characters are used
+    to jump to the correct line, and write over the line.
+    """
+
+    def __init__(self, stream, msg):
+        self.stream = stream
+        self.msg = msg
+        self.lines = []
+
+    def initialize(self, obj_index):
+        self.lines.append(obj_index)
+        self.stream.write("{} {} ... \r\n".format(self.msg, obj_index))
+        self.stream.flush()
+
+    def write(self, obj_index, status):
+        position = self.lines.index(obj_index)
+        diff = len(self.lines) - position
+        # move up
+        self.stream.write("%c[%dA" % (27, diff))
+        # erase
+        self.stream.write("%c[2K\r" % 27)
+        self.stream.write("{} {} ... {}\r".format(self.msg, obj_index, status))
+        # move back down
+        self.stream.write("%c[%dB" % (27, diff))
+        self.stream.flush()
```

Comment:
Making `write_out_msg()` a class allows us to:
* remove the need for the caller to be aware of `lines`, the state is managed within this class
* remove the need to pass around `stream` and the default message. It's always the same, so we can pass it in as a dependency to the constructor
* drop the if/else condition, they are now separate methods